### ISSUE TITLE
Unable to create a new region

### ISSUE THREAD

**Issue by Acedayz** [cyotek/Cyotek.Data.Nbt#2]
Hi, I've been trying to create a region (.mca) file from scratch, but I just can't get it to work.
I'm using "NBTExplorer" to read the file I'm creating for testing.

When I load the file in NBTExplorer, I can see the chunk that I'm creating but not any tags inside it.
So it seems that everything works fine, except the NBT part.

I'd appreciate any help with this issue.
Here's the class to create a region file:
`public class RegionFile
    {
        private const int SECTOR_BYTES = 4096;
        private const int SECTOR_INTS = SECTOR_BYTES / 4;
        private const int CHUNK_HEADER_SIZE = 5;
        private const int VERSION_DEFLATE = 2;

        private static byte[] emptySector = new byte[SECTOR_BYTES];
        private static int[] offsets;
        private static int[] timestamps;

        private List<bool> sectorFree;

        private static readonly DateTime Jan1st1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);

        public RegionFile(string path)
        {
            offsets = new int[SECTOR_INTS];
            timestamps = new int[SECTOR_INTS];

            try
            {
                using (BinaryWriter file = new BinaryWriter(File.Open(path, FileMode.Create)))
                {
                    if (file.BaseStream.Length < SECTOR_BYTES)
                    {
                        /* we need to write the chunk offset table */
                        for (int i = 0; i < SECTOR_INTS; ++i)
                        {
                            file.Write((int)0);
                        }
                        // write another sector for the timestamp info
                        for (int i = 0; i < SECTOR_INTS; ++i)
                        {
                            file.Write((int)0);
                        }
                    }

                    if ((file.BaseStream.Length & 0xfff) != 0)
                    {
                        /* the file size is not a multiple of 4KB, grow it */
                        for (int i = 0; i < (file.BaseStream.Length & 0xfff); ++i)
                        {
                            file.Write((byte)0);
                        }
                    }
                }

                using (BinaryReader file = new BinaryReader(File.Open(path, FileMode.Open)))
                {
                    /* set up the available sector map */
                    int nSectors = (int)file.BaseStream.Length / SECTOR_BYTES;
                    sectorFree = new List<bool>(nSectors);

                    for (int i = 0; i < nSectors; ++i)
                    {
                        sectorFree.Add(true);
                    }

                    sectorFree[0] = false; // chunk offset table
                    sectorFree[1] = false; // for the last modified info

                    file.BaseStream.Position = 0;
                    for (int i = 0; i < SECTOR_INTS; ++i)
                    {
                        int offset = file.ReadInt32();
                        offsets[i] = offset;
                        if (offset != 0 && (offset >> 8) + (offset & 0xFF) <= sectorFree.Count())
                        {
                            for (int sectorNum = 0; sectorNum < (offset & 0xFF); ++sectorNum)
                            {
                                sectorFree[(offset >> 8) + sectorNum] = false;
                            }
                        }
                    }
                    for (int i = 0; i < SECTOR_INTS; ++i)
                    {
                        int lastModValue = file.ReadInt32();
                        timestamps[i] = lastModValue;
                    }
                }
            }
            catch (IOException e)
            {
                MessageBox.Show(e.ToString());
            }
            
            // Tag Arrays
            byte[] biomeBytes = new byte[256];
            for (int i = 0; i < 256; ++i)
            {
                biomeBytes[i] = 0;
            }

            int[] heightMapInts = new int[256];
            for (int i = 0; i < 256; ++i)
            {
                heightMapInts[i] = 0;
            }

            byte[] blockBytes = new byte[4096];
            for (int i = 0; i < 4096; ++i)
            {
                blockBytes[i] = 56;
            }

            byte[] dataBytes = new byte[2048];
            for (int i = 0; i < 2048; ++i)
            {
                dataBytes[i] = 0;
            }

            byte[] blockLightBytes = new byte[2048];
            for (int i = 0; i < 2048; ++i)
            {
                blockLightBytes[i] = 0;
            }

            byte[] skyLightBytes = new byte[2048];
            for (int i = 0; i < 2048; ++i)
            {
                skyLightBytes[i] = 0;
            }
            // -

            TagCompound chunk = new TagCompound("Chunk");
            TagInt dataVersion = new TagInt("DataVersion", 176);
            TagCompound level = new TagCompound("Level");
            TagInt xPos = new TagInt("xPos", 0);
            TagInt zPos = new TagInt("zPos", 0);
            TagLong lastUpdate = new TagLong("LastUpdate", 0);
            TagByte lightPopulated = new TagByte("LightPopulated", 0);
            TagByte terrainPopulated = new TagByte("TerrainPopulated", 1);
            TagByte v = new TagByte("V", 1);
            TagLong inhabitedTime = new TagLong("InhabitedTime", 0);
            TagByteArray biomes = new TagByteArray("Biomes", biomeBytes);
            TagIntArray heightMap = new TagIntArray("HeightMap", heightMapInts);
            TagList sections = new TagList("Sections", TagType.Compound);
            TagCompound empty = new TagCompound();
            TagByte y = new TagByte("Y", 0);
            TagByteArray blocks = new TagByteArray("Blocks", blockBytes);
            TagByteArray data = new TagByteArray("Data", dataBytes);
            TagByteArray blockLight = new TagByteArray("BlockLight", blockLightBytes);
            TagByteArray skyLight = new TagByteArray("SkyLight", skyLightBytes);
            TagList entities = new TagList("Entities", TagType.Compound);
            TagList tileEntities = new TagList("TileEntities", TagType.Compound);

            dataVersion.Parent = chunk;
            level.Parent = chunk;
            xPos.Parent = level;
            zPos.Parent = level;
            lastUpdate.Parent = level;
            lightPopulated.Parent = level;
            terrainPopulated.Parent = level;
            v.Parent = level;
            inhabitedTime.Parent = level;
            biomes.Parent = level;
            heightMap.Parent = level;
            sections.Parent = level;
            empty.Parent = sections;
            y.Parent = empty;
            blocks.Parent = empty;
            data.Parent = empty;
            blockLight.Parent = empty;
            skyLight.Parent = empty;
            entities.Parent = level;
            tileEntities.Parent = level;

            NbtDocument doc = new NbtDocument(chunk);
            using (BinaryWriter file = new BinaryWriter(new FileStream(path, FileMode.Open)))
            {
                WriteChunk(file, 0, 0, doc, 258);
            }
        }

        void WriteChunk(BinaryWriter file, int x, int z, NbtDocument doc, int length)
        {
            try
            {
                int offset = getOffset(x, z);
                int sectorNumber = offset >> 8;
                int sectorsAllocated = offset & 0xFF;
                int sectorsNeeded = (length + CHUNK_HEADER_SIZE) / SECTOR_BYTES + 1;

                // Maximum chunk size is 1MB
                if (sectorsNeeded >= 256)
                {
                    return;
                }

                if (sectorNumber != 0 && sectorsAllocated == sectorsNeeded)
                {
                    WriteChunk(file, sectorNumber, doc, length);
                }
                else
                {
                    for (int i = 0; i < sectorsAllocated; ++i)
                    {
                        sectorFree[sectorNumber + i] = true;
                    }

                    int runStart = sectorFree.IndexOf(true);
                    int runLength = 0;
                    if (runStart != -1)
                    {
                        for (int i = runStart; i < sectorFree.Count(); ++i)
                        {
                            if (runLength != 0)
                            {
                                if (sectorFree[i])
                                    runLength++;
                                else
                                    runLength = 0;
                            }
                            else if (sectorFree[i])
                            {
                                runStart = i;
                                runLength = 1;
                            }
                            if (runLength >= sectorsNeeded)
                            {
                                break;
                            }
                        }
                    }

                    if (runLength >= sectorsNeeded)
                    {
                        /* We found a free space large enough */
                        sectorNumber = runStart;
                        SetOffset(file, x, z, (sectorNumber << 8) | sectorsNeeded);
                        for (int i = 0; i < sectorsNeeded; ++i)
                        {
                            sectorFree[sectorNumber + i] = false;
                        }
                        WriteChunk(file, sectorNumber, doc, length);
                    }
                    else {
                        // No free space large enough found; we need to grow the file

                        file.BaseStream.Position = (int)file.BaseStream.Length;
                        sectorNumber = sectorFree.Count;
                        for (int i = 0; i < sectorsNeeded; ++i)
                        {
                            file.Write(emptySector);
                            sectorFree.Add(false);
                        }
                        
                        WriteChunk(file, sectorNumber, doc, length);
                        SetOffset(file, x, z, (sectorNumber << 8) | sectorsNeeded);
                    }

                    SetTimestamp(file, x, z, CurrentTimeMillis());
                }
            }
            catch (IOException e)
            {
                MessageBox.Show(e.ToString());
            }
        }

        private void WriteChunk(BinaryWriter file, int sectorNumber, NbtDocument doc, int length)
        {
            file.BaseStream.Position = sectorNumber * SECTOR_BYTES;
            file.Write(length + 1); // Chunk length
            file.Write((byte)VERSION_DEFLATE); // Chunk version number
            doc.Save(file.BaseStream, CompressionOption.On);
        }

        int getOffset(int x, int z)
        {
            return offsets[x + z * 32];
        }

        void SetOffset(BinaryWriter file, int x, int z, int offset)
        {
            offsets[x + z * 32] = offset;
            file.BaseStream.Position = (x + z * 32) * 4;
            file.Write(offset);
        }

        void SetTimestamp(BinaryWriter file, int x, int z, int value)
        {
            timestamps[x + z * 32] = value;
            file.BaseStream.Position = SECTOR_BYTES + (x + z * 32) * 4;
            file.Write(value);
        }

        public static int CurrentTimeMillis()
        {
            return (int)(DateTime.UtcNow - Jan1st1970).TotalMilliseconds;
        }
    }`

I'd also like to know how I would go about getting the size of the NbtDocument (how many bytes it saves).

**Comment by cyotek** [comment]
Hello,

The only thing I noted in your example code is that when writing your chunk data you're using the value `VERSION_DEFLATE`. The current version of the NBT library doesn't allow you to specify *which* compression method to use - so while it can **read** uncompressed, deflate compressed and gzip compressed, it always **writes** using gzip compression (when compression is enabled). Try using a value of `1` instead to tell NBTExplorer etc to read in gzip format instead.

While I'm not familiar with the region format (I use this library for my own data files, not Minecraft's), there is actually a [test](https://github.com/cyotek/Cyotek.Data.Nbt/blob/master/src/Cyotek.Data.Nbt.Tests/AnvilRegionTests.cs) from the original author of the code which is pretty much untouched and seems to be dealing with similar constructs as your sample - that seems to bear out my theory of the compression flag.

Finally, don't forget the NBT format requires values to be stored in big endian format - and judging from the test I mentioned, so does the Anvil/MCA format, so if you're working on Windows you'll need to do some byte swapping.

As for getting the size of the document, that's a tricky one. I could add something to give you the **uncompressed** size, but not the **compressed** size. However, as you are working directly with streams you can probably work it out yourself - just store the current `Stream.Position}} before saving, then subtract the final position from this value to get the number of bytes read.

Sorry I can't give you a detailed answer, but as I mentioned while I used to play Minecraft an awful lot, that's not what I use this library for. Hopefully that test from the original author will give you some insight though!

Regards;
Richard Moss

**Comment by Acedayz** [comment]
Thanks, that answers every question I had.
I've never done any byte swapping before, and that's what I thought was needed as well. I guess I'll have to figure out how that works.
I just hope that Minecraft or NBTExplorer can read it in gzip format and maybe convert it to zlib format(?).
If it ends up not working and I need to directly write it using zlib compression, is there anything I can do about it with your library (nothing too advanced since I'm not an expert)?

**Comment by cyotek** [comment]
I think Minecraft just uses deflate or gzip so writing the correct value into your region file should be enough (assuming you resolve any niggles regarding little-endian/big-engian) - as the signature bytes are different, I assume NBTExplorer is failing to decompress the data and therefore silently ignoring it. But I'm just guessing, I don't know for sure (unhelpful I know)

There isn't any way currently of forcing the compression method - when I recently overhauled the library to allow streams to be directly used (the previous version only allowed for filenames) I did briefly think about it, but decided it was just adding another level of complexity. Perhaps I should update it to add another overload to allow the compression method to be specified for the binary format.
